pkg/kubectl/cmd/util/factory.go

```diff
@@ -554,21 +593,21 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 	}
 }
 
-// GetFirstPod returns the first pod of an object from its namespace and selector
-func GetFirstPod(client *client.Client, namespace string, selector labels.Selector) (*api.Pod, error) {
+// GetFirstPod returns the first pod of an object from its namespace and selector and the number of matching pods
+func GetFirstPod(client *client.Client, namespace string, selector labels.Selector) (*api.Pod, int, error) {
 	var pods *api.PodList
 	for pods == nil || len(pods.Items) == 0 {
 		var err error
 		options := api.ListOptions{LabelSelector: selector}
 		if pods, err = client.Pods(namespace).List(options); err != nil {
-			return nil, err
+			return nil, 0, err
 		}
 		if len(pods.Items) == 0 {
 			time.Sleep(2 * time.Second)
 		}
 	}
 	pod := &pods.Items[0]
```

Comment:
> Shouldn't we sort this list and pick the oldest pod?

I didn't want to change the order on the other callers.  If `kubectl logs` chooses one, the oldest one isn't likely to be the correct choice. You probably want to check things like status to find the most recent one that failed.  For attach, you want to find one that's running and probably the most recent one.

At any rate, I'm not convinced that sorting makes the behavior better.